Script to convert Grok-1 weights from raw JAX pickle files.

heiner opened the pull request:
This adds a script to convert the [raw weights in the pickle files](https://github.com/xai-org/grok-1?tab=readme-ov-file#downloading-the-weights) to GGUF format. This allows using @arki05's work in #6204 directly from the Grok-1 torrent.

Code is based on @foldl's conversion script in [chatllm.cpp](https://github.com/foldl/chatllm.cpp/blob/master/convert.py), which in turn is based on @chu-tianxiang's [gist](https://gist.github.com/chu-tianxiang/ec310e15d56949fd0f351cb5f65ee7a1).

Main ideas to avoid excessive memory:

* Parse pickle files using `mmap`.
* Use PyTorch "meta" tensors to simulate shape and dtype results without having to do all conversions beforehand.

Note that I couldn't run the full model due to RAM constrains and it's possible I mixed up some tensor names.

heiner committed:
Script to convert Grok-1 weights from raw JAX pickle files.

slaren commented:
Does this merge the experts into a single tensor?

heiner commented:
> Does this merge the experts into a single tensor?

It does the opposite -- in the raw data, the 8 experts are part of the same tensor. This splits them, which is also chat the the [chatllm.ccp script](https://github.com/foldl/chatllm.cpp/blob/master/convert.py) does.

If there is a way to keep them within one tensor I'm happy to make that change.

heiner commented:
Understood. That will actually make the script simpler. Would you happen to know the tensor names I should use in this case? Currently when using splitting, they are

```
| blk.{layer}.ffn_gate.{expert}.weight        | torch.Size([32768, 6144])  | Q4_0    |
| blk.{layer}.ffn_down.{expert}.weight        | torch.Size([6144, 32768])  | Q4_0    |
| blk.{layer}.ffn_up.{expert}.weight          | torch.Size([32768, 6144])  | Q4_0    |
```

slaren commented:
The tensor names are defined in gguf-py:
https://github.com/ggerganov/llama.cpp/blob/60325fa56f61c228464c9f065db3aa6a61f2156e/gguf-py/gguf/constants.py#L249-L251

It would be good to use these constants rather than hardcoding the names.

heiner committed:
Don't split MoE weights.

As per https://github.com/ggerganov/llama.cpp/pull/7058#issuecomment-2092967508.
This helps avoid a memcopy when running.

heiner committed:
Use only one list of weight names, with values from the gguf module.

This saves weights in the order in which they are in the Grok-1 files.
Since we operate weight-by-weight now, we no longer need caches and
name2key translations.

Per reviewer request, I also moved to using keys in gguf.TENSOR_NAMES.

heiner commented:
Thanks!

I have updated the branch to no longer split the MoE weights into separate tensors. That simplifies the script as it's now one weight per file. The original script permutated the order in which these weights are written for some reason, I stopped doing that now and thus there's only one list of weight names.

I also moved to the values in the `gguf.TENSOR_NAMES` dict as per your suggestion. I'm not sure that's a clear improvement ("Explicit is better than implicit."), especially in view of code like `name.endswith("attn_k")` and `name.endswith("_exps")` but it's also not much worse.

PTAL.

foldl commented:
@heiner, name of my project is `ChatLLM.cpp`, not `ChatLLM.ccp`, 😄

ggerganov reviewed (dismissed):
We can merge after lint fixes

ggerganov commented:
Hm, I tested `Q4_0` conversion and it does not seem to work:

```
python3 convert_grok.py -i ~/Data/huggingface/grok-1/ckpt/ --vocab_dir ../grok-1 -o x.gguf -t q4_0

make -j && ./main -m x.gguf -p "I believe the meaning of life is" -ngl 0

...

[BOS] I believe the meaning of life is it:000000000000000 of0000000000000000000000000000000000000000000000

```

Might need more work

foldl commented on convert_grok.py:
```diff
@@ -0,0 +1,458 @@
+            .reshape(weights.shape)
+        )
+
+    if name.endswith("attn_k"):
+        return permute(tensor, config.num_key_value_heads)
+    elif name.endswith("attn_q"):
+        return permute(tensor, config.num_attention_heads)
```
Grok is using `LLAMA_ROPE_TYPE_NEOX`, so do not `permute`.

heiner replied:
Removed the call to this function and the function itself.

foldl commented on convert_grok.py:
```diff
@@ -0,0 +1,458 @@
+        "pad_token_id": 0,
+        "eos_token_id": 2,
+        "max_position_embeddings": 8192,
+        "output_multiplier_scale": 0.5773502691896257,
+        "embedding_multiplier_scale": 78.38367176906169,
```
`embedding_multiplier_scale` should be multiplied to embeddings. 

`output_multiplier_scale` can be ignored, IMO.

heiner replied:
Added

```python
    if name == "token_embd":
        weight *= config.embedding_multiplier_scale
```

heiner replied:
My original reason for not doing that was that this appears to also happen in the C++ code here: https://github.com/ggerganov/llama.cpp/blob/befddd0f15de6efb15d7e7f5b527dfb671f4196f/llama.cpp#L7487

foldl replied:
Oh, I missed this. Then, `embedding_multiplier_scale` should not be multiplied.

But, it's better to do the multiplication once and offline than to do it each time at runtime.

heiner commented:
```
[BOS] I believe the meaning of life is it:000000000000000 of0000000000000000000000000000000000000000000000
```

My apologies. As I said above I couldn't actually test running the full model on my setup. I will fix @foldl's suggestions.

Would you happen to have something like the sha-1 of each tensor of a checkpoint based on the HF weights? Otherwise I can download those and run that conversion for comparision.

heiner committed:
Move print to logging: Fixes.

mofosyne committed:
Update convert_grok.py to use logging module

heiner committed:
Address review comments by foldl.

heiner commented:
Thanks @foldl for the hints. It's well possible I mixed something else up as well, e.g., swapped two tensors with the same shape and dtype. Would you happen to have a `tensor name -> hash` table for a correct conversion?

foldl commented:
@heiner You need to compare the result against https://huggingface.co/keyfan/grok-1-hf, although I don't think sha-1 could match.

chatllm.cpp `permute`s k_proj/q_proj weights, so, sha-1 would not match, either.

heiner committed:
Don't multiply embeddings with embedding_multiplier_scale as it happens in llama.cpp.

heiner commented:
Thanks. I have removed the multiplication with `embedding_multiplier_scale` again and converted the full model with all 8 experts. The output is not great but also not as bad as before ([gist with full output](https://gist.github.com/heiner/1e036f5714718e404392b384bd0e2ac3)):

```
./build/bin/main -m grok.bin -p "I believe the meaning of life is" -s 2 -n 10 -ngl 0
(...)
[BOS] I believe the meaning of life is important it could possibly the general of the general I
```

It's likely something else is wrong but I'm unsure what it is, and the multiple-hour iteration time makes it infeasible to just try out random things.

heiner committed:
Fix layer order.

heiner committed:
Use Q8_0 quantization from gguf module.

This makes tensors exactly as in https://huggingface.co/Arki05/Grok-1-GGUF/tree/main/Q8_0

heiner committed:
More constants from gguf.

heiner committed:
Write tensors in layer order.

heiner committed:
Move noqa comment to where the lastest flake8 likes it.

heiner commented on convert_grok.py:
```diff
@@ -0,0 +1,460 @@
+    assert len(arrays) in (1, 2)
+
+
+def quantize_q8_0(tensor: torch.Tensor) -> torch.CharTensor:
+    # equivalent to ggml_quantize_q8_0 in ggml.c (modulo rounding away from zero)
+    assert tensor.shape[1] % QK8_0 == 0
+    tensor = tensor.reshape(-1, QK8_0)
+    scale = tensor.abs().max(dim=-1, keepdim=True).values / ((1 << 7) - 1)
+    tensor = (tensor / scale).round().clamp(min=-128, max=127).char()
+    # add scale into each block
+    tensor = torch.cat((scale.half().view(torch.int8), tensor), dim=-1)
+    return tensor
```
I'm happy to add that. I tried adding custom rounding but I didn't try the second point at the same time.

heiner replied:
Done. I've verified the weights are still the same as in [Arki05/Grok-1-GGUF/Q8_0](https://huggingface.co/Arki05/Grok-1-GGUF/tree/main/Q8_0).

heiner committed:
Implement Q8_0 quantization fully in PyTorch.

This is equivalent to gguf.quantize_q8_0 but doesn't round-trip to
Numpy.

foldl commented:
> The Grok-1 pickle files use lexicographic order ...

Nice catch, 😄

heiner commented:
(The Docker image test failure is unrelated to the changes in this PR.)

mofosyne commented:
> (The Docker image test failure is unrelated to the changes in this PR.)

Yup, it's bypassed for now. Please rebase on top of current master

heiner commented:
> > (The Docker image test failure is unrelated to the changes in this PR.)
> 
> Yup, it's bypassed for now. Please rebase on top of current master

Done.

ggerganov commented:
Does the conversion work correctly now? I can run some tests if you need confirmation?

heiner commented:
The actual generations still don't look great, e.g. for Q8_0:

```
$ ./build/bin/main -m grok.bin -p "The answer to life the universe and everything is" -s 1 -n 4 -ngl 2
...
[BOS] The answer to life the universe and everything is gifted for the of
```

What would be very useful is if you could run both this (in Q8_0) as well as [Arki05/Grok-1-GGUF/Q8_0](https://huggingface.co/Arki05/Grok-1-GGUF/tree/main/Q8_0) and let me know if there's any difference in the output, and if so, what plausibly could cause that. I have verified that the actual weights in both cases are the same, so any difference would presumably be kv settings.
